src/amber/pipes/cors.cr

```diff
@@ -1,81 +1,133 @@
-        end
+      private def put_response_headers(response)
+        response.headers[Headers::ALLOW_CREDENTIALS] = @credentials.to_s if @credentials
+        response.headers[Headers::ALLOW_ORIGIN] = @origin.request_origin.not_nil!
+        response.headers[Headers::VARY] = vary unless @origin.any?
+      end
 
-        if max_age > 0
-          context.response.headers["Access-Control-Max-Age"] = max_age.to_s
+      private def vary
+        String.build do |str|
+          str << Headers::ORIGIN
+          str << "," << @vary if @vary
         end
+      end
+    end
 
-        # if asking permission for request method or request headers
-        if context.request.method.downcase == "options"
-          context.response.status_code = 200
-          response = ""
-
-          if requested_method = context.request.headers["Access-Control-Request-Method"]
-            if allow_methods.includes? requested_method.strip
-              context.response.headers["Access-Control-Allow-Methods"] = allow_methods.join(", ")
-            else
-              context.response.status_code = 403
-              response = "Method #{requested_method} not allowed."
-            end
-          end
+    module Preflight
+      extend self
 
-          if requested_headers = context.request.headers["Access-Control-Request-Headers"]
-            requested_headers.split(",").each do |requested_header|
-              if allow_headers.includes? requested_header.strip.downcase
-                context.response.headers["Access-Control-Allow-Headers"] = allow_headers.join(", ")
-              else
-                context.response.status_code = 403
-                response = "Headers #{requested_headers} not allowed."
-              end
-            end
+      def request?(context, cors)
+        if context.request.method == "OPTIONS"
+          if valid_method?(context.request, cors.methods) &&
+             valid_headers?(context.request, cors.headers)
+            set_preflight_headers(context.request, context.response, cors.max_age)
+          else
+            cors.forbidden(context)
           end
+        end
+      end
 
-          context.response.content_type = "text/html; charset=utf-8"
-          context.response.print(response)
-        else
-          call_next(context)
+      def set_preflight_headers(request, response, max_age)
+        response.headers[Headers::ALLOW_METHOD] = request.headers[Headers::REQUEST_METHOD]
+        response.headers[Headers::ALLOW_HEADERS] = request.headers[Headers::REQUEST_HEADERS]
+        response.headers[Headers::ALLOW_MAX_AGE] = max_age.to_s if max_age
+        response.content_length = 0
+        response.flush
+      end
+
+      def valid_method?(request, methods)
+        methods.includes? request.headers[Headers::REQUEST_METHOD]?
+      end
+
+      def valid_headers?(request, headers)
+        !(headers & request.headers[Headers::REQUEST_HEADERS].split(',')).empty?
+      end
+    end
+
+    struct Origin
```

Comment:
Above `OriginType` is defined, but it's not used at all in this struct. Should it be?

Reply:
This was updated